Sources/PackageGraph/RegistryPackageContainer.swift

```diff
@@ -0,0 +1,204 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import PackageLoading
+import PackageModel
+import PackageRegistry
+import SourceControl
+
+import TSCBasic
+import TSCUtility
+
+import Dispatch
+
+public final class RegistryPackageContainer {
+    public var identifier: PackageReference
+
+    let mirrors: DependencyMirrors
+
+    /// The manifest loader.
+    let manifestLoader: ManifestLoaderProtocol
+
+    /// The tools version loader.
+    let toolsVersionLoader: ToolsVersionLoaderProtocol
+
+    /// The current tools version in use.
+    let currentToolsVersion: ToolsVersion
+
+    let registryManager: RegistryManager
+
+    private(set) var reversedVersionsCache: [Version] = []
+    private var manifestsCache = ThreadSafeKeyValueStore<Version, Manifest>()
+    private var toolsVersionsCache = ThreadSafeKeyValueStore<Version, ToolsVersion>()
+    private var validToolsVersionsCache = ThreadSafeKeyValueStore<Version, Bool>()
+
+    private init(
+        _ identifier: PackageReference,
+        mirrors: DependencyMirrors,
+        registryManager: RegistryManager,
+        manifestLoader: ManifestLoaderProtocol,
+        toolsVersionLoader: ToolsVersionLoaderProtocol,
+        currentToolsVersion: ToolsVersion
+    ) {
+        self.identifier = identifier
+        self.mirrors = mirrors
+        self.registryManager = registryManager
+        self.manifestLoader = manifestLoader
+        self.toolsVersionLoader = toolsVersionLoader
+        self.currentToolsVersion = currentToolsVersion
+    }
+
+    static func create(
+        for identifier: PackageReference,
+        mirrors: DependencyMirrors,
+        registryManager: RegistryManager,
+        manifestLoader: ManifestLoaderProtocol,
+        toolsVersionLoader: ToolsVersionLoaderProtocol,
+        currentToolsVersion: ToolsVersion,
+        completion: (Result<RegistryPackageContainer, Error>) -> Void
+    ) {
+        let container = RegistryPackageContainer(identifier,
+                                                 mirrors: mirrors,
+                                                 registryManager: registryManager,
+                                                 manifestLoader: manifestLoader,
+                                                 toolsVersionLoader: toolsVersionLoader,
+                                                 currentToolsVersion: currentToolsVersion)
+
+        do {
+            let versionsWithDuplicates = try tsc_await { registryManager.fetchVersions(of: identifier, completion: $0) }
```

Comment:
since the function is async no need to await here, instead handle the completion